### PR TITLE
2889-support-unlinking-research-data-from-user

### PR DESCRIPTION
**Pull request by afs25**
Ready for review; all tests passing both locally and on AWS. Requires relevant migrations to be applied.

### CODE REVIEW

File: api/endpoints/user_task.py [hunk 1]
```diff
@@ -252,10 +258,11 @@ def create_user_task(ut_json, correlation_id):
...
             project_task_id,
             status,
-            consented
-        ) VALUES ( %s, %s, %s, %s, %s, %s, %s );'''
+            consented,
+            ext_user_task_id
+        ) VALUES ( %s, %s, %s, %s, %s, %s, %s, %s );'''
 
-    execute_non_query(sql, (id, created, created, user_project_id, project_task_id, ut_status, ut_consented), correlation_id)
+    execute_non_query(sql, (id, created, created, user_project_id, project_task_id, status, ut_consented, ext_user_task_id), correlation_id)
```
Comment (this-andy) [inline api/endpoints/user_task.py]: Remove ut_ from consented too

File: api/endpoints/user_task.py [hunk 2]
```diff
@@ -195,6 +204,21 @@ def create_user_task(ut_json, correlation_id):
         raise DetailedValueError('mandatory data missing', errorjson) from err
 
     # now process optional json data
+    for variable_name, default_value, validating_func in [
+        ('ext_user_task_id', str(uuid.uuid4()), validate_uuid),
+        ('created', str(now_with_tz()), validate_utc_datetime),
+        ('status', DEFAULT_STATUS, validate_status),
```
Comment (this-andy) [inline api/endpoints/user_task.py]: Extract this as a variable

File: api/endpoints/user_project.py [hunk 3]
```diff
@@ -138,6 +146,21 @@ def create_user_project(up_json, correlation_id, do_nothing_if_exists=False):
         raise DetailedValueError('mandatory data missing', errorjson) from err
 
     # now process optional json data
+    for variable_name, default_value, validating_func in [
+        ('ext_user_project_id', str(uuid.uuid4()), validate_uuid),
+        ('created', str(now_with_tz()), validate_utc_datetime),
+        ('status', DEFAULT_STATUS, validate_status),
```
Comment (this-andy) [inline api/endpoints/user_project.py]: Extract as variable

File: api/local/database-view-sql/view_external_users_identity.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        u.id AS user_id,
+        u.email,
+        u.first_name,
+        u.last_name,
+        ut.id AS user_task_id,
+        ut.ext_user_task_id,
+        up.id AS user_project_id,
+        up.ext_user_project_id,
+        pt.id AS project_task_id,
+        p.id AS project_id
```
Comment (this-andy) [inline api/local/database-view-sql/view_external_users_identity.sql]: Can we add project.name, project_task.description

File: api/tests/test_scripts/test_user.py [hunk 5]
```diff
@@ -41,32 +45,45 @@
...
     def setUpClass(self):
         set_running_unit_tests(True)
 
-        truncate_table('public.projects_user')
-        truncate_table('public.projects_entityupdate')
+        pg_utils.truncate_table_multiple(
+            'public.projects_usergroup',
+            'public.projects_project',
+            'public.projects_user',
+            'public.projects_userproject',
```
Comment (this-andy) [inline api/tests/test_scripts/test_user.py]: add entityupdate table here, and in teardown

**Comment by afs25** [comment]
Addressed all requested changes. All tests passing locally. This is now ready for review/merge.

**Pull request merged**